### ISSUE TITLE
failure-v0.1.5 fails to compile with rustc-1.37.0

### ISSUE THREAD

**Issue by stefson** [rust-lang-nursery/failure#324]
there's a ton of failures with rustc-1.37.0, while it was working with rustc-1.34.2: 

```
cd /tmp/failure && cargo build
   Compiling proc-macro2 v0.4.30
   Compiling unicode-xid v0.1.0
   Compiling libc v0.2.62
   Compiling syn v0.15.44
   Compiling cc v1.0.45
   Compiling failure_derive v0.1.5 (/tmp/failure/failure_derive)
   Compiling rustc-demangle v0.1.16
   Compiling cfg-if v0.1.9
   Compiling quote v0.6.13
   Compiling backtrace-sys v0.1.31
   Compiling backtrace v0.3.37
   Compiling synstructure v0.10.2
   Compiling failure v0.1.5 (/tmp/failure)
error: trait objects without an explicit `dyn` are deprecated
  --> src/as_fail.rs:11:27
   |
11 |     fn as_fail(&self) -> &Fail;
   |                           ^^^^ help: use `dyn`: `dyn Fail`
   |
note: lint level defined here
  --> src/lib.rs:25:9
   |
25 | #![deny(warnings)]
   |         ^^^^^^^^
   = note: #[deny(bare_trait_objects)] implied by #[deny(warnings)]

error: trait objects without an explicit `dyn` are deprecated
  --> src/as_fail.rs:18:27
   |
18 |     fn as_fail(&self) -> &Fail {
   |                           ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
  --> src/as_fail.rs:23:17
   |
23 | impl AsFail for Fail {
   |                 ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
  --> src/as_fail.rs:24:27
   |
24 |     fn as_fail(&self) -> &Fail {
   |                           ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
  --> src/as_fail.rs:33:31
   |
33 |         fn as_fail(&self) -> &Fail {
   |                               ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
 --> src/box_std.rs:5:27
  |
5 | pub struct BoxStd(pub Box<Error + Send + Sync + 'static>);
  |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use `dyn`: `dyn Error + Send + Sync + 'static`

error: trait objects without an explicit `dyn` are deprecated
  --> src/compat.rs:42:30
   |
42 |     impl From<Error> for Box<StdError> {
   |                              ^^^^^^^^ help: use `dyn`: `dyn StdError`

error: trait objects without an explicit `dyn` are deprecated
  --> src/compat.rs:43:38
   |
43 |         fn from(error: Error) -> Box<StdError> {
   |                                      ^^^^^^^^ help: use `dyn`: `dyn StdError`

error: trait objects without an explicit `dyn` are deprecated
  --> src/compat.rs:48:30
   |
48 |     impl From<Error> for Box<StdError + Send + Sync> {
   |                              ^^^^^^^^^^^^^^^^^^^^^^ help: use `dyn`: `dyn StdError + Send + Sync`

error: trait objects without an explicit `dyn` are deprecated
  --> src/compat.rs:49:38
   |
49 |         fn from(error: Error) -> Box<StdError + Send + Sync> {
   |                                      ^^^^^^^^^^^^^^^^^^^^^^ help: use `dyn`: `dyn StdError + Send + Sync`

error: trait objects without an explicit `dyn` are deprecated
   --> src/context.rs:115:36
    |
115 |         fn cause(&self) -> Option<&Fail> {
    |                                    ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/context.rs:149:39
    |
149 |         fn as_cause(&self) -> Option<&Fail> {
    |                                       ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
 --> src/error/error_impl.rs:7:22
  |
7 |     inner: Box<Inner<Fail>>,
  |                      ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
  --> src/error/error_impl.rs:28:38
   |
28 |     pub(crate) fn failure(&self) -> &Fail {
   |                                      ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
  --> src/error/error_impl.rs:32:50
   |
32 |     pub(crate) fn failure_mut(&mut self) -> &mut Fail {
   |                                                  ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
  --> src/error/mod.rs:63:39
   |
63 |     pub fn from_boxed_compat(err: Box<StdError + Sync + Send + 'static>) -> Error {
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use `dyn`: `dyn StdError + Sync + Send + 'static`

error: trait objects without an explicit `dyn` are deprecated
  --> src/error/mod.rs:69:31
   |
69 |     pub fn as_fail(&self) -> &Fail {
   |                               ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
  --> src/error/mod.rs:85:29
   |
85 |     pub fn cause(&self) -> &Fail {
   |                             ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/error/mod.rs:136:39
    |
136 |     pub fn find_root_cause(&self) -> &Fail {
    |                                       ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/error/mod.rs:176:34
    |
176 |     pub fn root_cause(&self) -> &Fail {
    |                                  ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/error/mod.rs:204:12
    |
204 | impl AsRef<Fail> for Error {
    |            ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/error/mod.rs:205:26
    |
205 |     fn as_ref(&self) -> &Fail {
    |                          ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:129:32
    |
129 |     fn cause(&self) -> Option<&Fail> {
    |                                ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:189:30
    |
189 |     fn root_cause(&self) -> &Fail
    |                              ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:202:6
    |
202 | impl Fail {
    |      ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:208:45
    |
208 |             unsafe { Some(&*(self as *const Fail as *const T)) }
    |                                             ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:220:47
    |
220 |             unsafe { Some(&mut *(self as *mut Fail as *mut T)) }
    |                                               ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:234:39
    |
234 |     pub fn find_root_cause(&self) -> &Fail {
    |                                       ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:261:34
    |
261 |     pub fn root_cause(&self) -> &Fail {
    |                                  ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:276:19
    |
276 | impl Fail for Box<Fail> {
    |                   ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:277:32
    |
277 |     fn cause(&self) -> Option<&Fail> {
    |                                ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:288:22
    |
288 |     fail: Option<&'f Fail>,
    |                      ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:292:21
    |
292 |     type Item = &'f Fail;
    |                     ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:293:38
    |
293 |     fn next(&mut self) -> Option<&'f Fail> {
    |                                      ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:301:31
    |
301 | fn find_root_cause(mut fail: &Fail) -> &Fail {
    |                               ^^^^ help: use `dyn`: `dyn Fail`

error: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:301:41
    |
301 | fn find_root_cause(mut fail: &Fail) -> &Fail {
    |                                         ^^^^ help: use `dyn`: `dyn Fail`

error: aborting due to 36 previous errors

error: Could not compile `failure`.

To learn more, run the command again with --verbose.
```

p.s: I used cargo build to compile

**Comment by ArekPiekarz** [comment]
Are you treating warnings as errors in your project? Because `trait objects without an explicit 'dyn' are deprecated` is by default emitted as a warning.

**Comment by stefson** [comment]
I initially noticed this during firefox compile, where they are treated as warnings, yes. The log parsed in my posting is from the git checkout of `failure` - it might be, that my rust install does turn warnings into errors by default, but still this behavior seems to be deprecated and the warning valid, right?

**Comment by ArekPiekarz** [comment]
I don't think it changes behavior, it just improves readability - with `dyn` you now know a dynamic dispatch will be used instead of static.

**Comment by Ixrec** [comment]
The "bare trait" **syntax** is deprecated in favor of `dyn` syntax, but yes, the meaning and runtime behavior is identical.

**Comment by Ixrec** [comment]
And yes the warnings are entirely correct, but "fails to compile" is an incorrect description of the issue, since you have to opt-in to "warnings are errors" behavior (after all, if it wasn't opt-in, we'd never be allowed to add new warnings).

**Comment by davidbarsky** [comment]
I've published 0.1.6 that should fix these issues.
